src/dotty/tools/dotc/core/pickling/UnPickler.scala

```diff
@@ -118,6 +118,13 @@ object UnPickler {
         denot.owner.thisType select denot.sourceModule
       else selfInfo
     if (!(denot.flagsUNSAFE is JavaModule)) ensureConstructor(denot.symbol.asClass, decls)
+    if (denot.flagsUNSAFE is Module) {
+      val scalacCompanion = denot.classSymbol.scalacLinkedClass
+      val companionClassMethod = ctx.synthesizeCompanionMethod(nme.COMPANION_CLASS_METHOD, scalacCompanion, denot.classSymbol)
+      if (companionClassMethod.exists)
+        companionClassMethod.entered
+    }
+
```

Comment:
It's a pity that we still need scalacLinkedClass here, because it means we cannot get rid of the associated complexity of isCoDefinedWith. I would much prefer if we got the linked class directly from the unpickling info, instead of looking at 
the scope.